### PR TITLE
Deep-merge-upstream-annotation-config

### PR DESCRIPTION
**Pull request by ben-taussig-solo**
# Description
Currently the feature that allows customizing discovered upstreams via the gloo.solo.io/upstream_config annotation (https://github.com/solo-io/gloo/issues/3717 ) only supports overwriting top-level upstream fields.

This PR adds the ability to deep merge discovered annotation config specified via the `gloo.solo.io/upstream_config` annotation with default upstream config. A user can enable this feature by setting the `gloo.solo.io/upstream_config/deep_merge` annotation to `true` on the service for which an upstream is discovered.

This PR uses the reflection based library [mergo](https://github.com/imdario/mergo) to merge upstreams. The library is used in a significant number of high-profile projects, so I am personally comfortable using it in Gloo Edge.

BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/6005

### CODE REVIEW

File: docs/content/guides/traffic_management/destination_types/discovered_upstream/discovered-upstream-configuration.md [hunk 1]
```diff
@@ -98,4 +98,10 @@ status:
       state: 1
 {{< /highlight >}}
 
-As you can see, the configuration set `spec.initialStreamWindowSize` to `2048` on the discovered upstream! 
\ No newline at end of file
+As you can see, the configuration set `spec.initialStreamWindowSize` to `2048` on the discovered upstream! 
+
+## Merge strategies
+
+By default, discovered upstreams configured via the gloo.solo.io/upstream_config annotation will completely overwrite top-level upstream fields for which configuration has been specified.
```
Comment (jenshu) [inline docs/content/guides/traffic_management/destination_types/discovered_upstream/discovered-upstream-configuration.md:105]:
```suggestion
By default, discovered upstreams configured via the `gloo.solo.io/upstream_config` annotation will completely overwrite top-level upstream fields for which configuration has been specified.
```

File: projects/gloo/pkg/plugins/kubernetes/uds_convert_test.go [hunk 2]
```diff
@@ -154,6 +159,176 @@ var _ = Describe("UdsConvert", func() {
 				Expect(upstreamConfigJson).To(Equal(expectedCfgJson))
 			}
 
+			CreateUpstreamWithSpec := func(uc *KubeUpstreamConverter, ctx context.Context, svc *kubev1.Service, port kubev1.ServicePort, upstream *v1.Upstream) *v1.Upstream {
+				for _, sc := range uc.serviceConverters {
+					if err := sc.ConvertService(svc, port, upstream); err != nil {
+						contextutils.LoggerFrom(ctx).Errorf("error: failed to process service options with err %v", err)
```
Comment (jenshu) [inline projects/gloo/pkg/plugins/kubernetes/uds_convert_test.go:165]: can we just return an error here?

File: projects/gloo/pkg/plugins/kubernetes/uds_convert_test.go [hunk 3]
```diff
@@ -154,6 +159,176 @@ var _ = Describe("UdsConvert", func() {
...
+					coremeta *core.Metadata
+					labels   map[string]string
+				)
+
+				BeforeEach(func() {
+					svc = &kubev1.Service{
+						Spec: kubev1.ServiceSpec{},
+					}
+					svc.Name = "test"
+					svc.Namespace = "test"
```
Comment (jenshu) [inline projects/gloo/pkg/plugins/kubernetes/uds_convert_test.go:185]: super-nit, can we have name and namespace be different?

File: projects/gloo/pkg/plugins/kubernetes/uds_convert_test.go [hunk 4]
```diff
@@ -154,6 +159,176 @@ var _ = Describe("UdsConvert", func() {
...
+					coremeta.ResourceVersion = ""
+					coremeta.Name = UpstreamName(meta.Namespace, meta.Name, port.Port)
+					labels = coremeta.GetLabels()
+					coremeta.Labels = make(map[string]string)
+				})
+				It("should merge upstream fields properly with merge enabled", func() {
+					annotations := map[string]string{
+						serviceconverter.DeepMergeAnnotationPrefix: "true",
+						serviceconverter.GlooAnnotationPrefix: `{
+							"ssl_config": {
```
Comment (jenshu) [inline projects/gloo/pkg/plugins/kubernetes/uds_convert_test.go:202]: should these be in camelCase? (same with other tests below)

File: projects/gloo/pkg/plugins/kubernetes/uds_convert_test.go (hunk 2 above)
Comment (ben-taussig-solo) [reply projects/gloo/pkg/plugins/kubernetes/uds_convert_test.go]: Resolved here: https://github.com/solo-io/gloo/pull/6036/commits/fd8078fe46999a8b0344b63828a0c0f2cdea47cc

File: projects/gloo/pkg/plugins/kubernetes/uds_convert_test.go (hunk 4 above)
Comment (ben-taussig-solo) [reply projects/gloo/pkg/plugins/kubernetes/uds_convert_test.go:202]: Ah yes, they should. The feature is functional using snake_case but we're trying to direct users to use camelCase.

File: projects/gloo/pkg/plugins/kubernetes/uds_convert_test.go (hunk 3 above)
Comment (ben-taussig-solo) [reply projects/gloo/pkg/plugins/kubernetes/uds_convert_test.go:185]: sure!

File: projects/gloo/pkg/plugins/kubernetes/uds_convert_test.go (hunk 4 above)
Comment (ben-taussig-solo) [reply projects/gloo/pkg/plugins/kubernetes/uds_convert_test.go]: Resolved here: https://github.com/solo-io/gloo/pull/6036/commits/d8d71738b477eaf78f8406960995b55e191cc779

File: projects/gloo/pkg/plugins/kubernetes/uds_convert_test.go (hunk 3 above)
Comment (ben-taussig-solo) [reply projects/gloo/pkg/plugins/kubernetes/uds_convert_test.go]: Resolved here: https://github.com/solo-io/gloo/pull/6036/commits/d8d71738b477eaf78f8406960995b55e191cc779

File: projects/gloo/pkg/plugins/kubernetes/uds_convert_test.go [hunk 5]
```diff
@@ -154,6 +158,178 @@ var _ = Describe("UdsConvert", func() {
...
+				for _, sc := range uc.serviceConverters {
+					if err := sc.ConvertService(svc, port, upstream); err != nil {
+						return nil, err
+					}
+				}
+
+				return upstream, nil
+			}
+
+			Describe("deep merge", func() {
```
Comment (jenshu) [inline projects/gloo/pkg/plugins/kubernetes/uds_convert_test.go:171]: could you also add a test with DeepMergeAnnotationPrefix=false and make sure it retains the old behavior?

Comment (jenshu) [reply projects/gloo/pkg/plugins/kubernetes/uds_convert_test.go:171]: otherwise, LGTM!

Comment (ben-taussig-solo) [reply projects/gloo/pkg/plugins/kubernetes/uds_convert_test.go:171]: Resolved here: https://github.com/solo-io/gloo/pull/6036/commits/3148630610be63c45397cc87fc83621532bf7e77
thanks for reviewing!!

**Review by jenshu — APPROVED** [review]

**Pull request merged**